src/modrover/rover.py

```diff
@@ -113,6 +169,101 @@ def explore(self, dataset: pd.DataFrame, strategy: Union[str, RoverStrategy]):
         """Fit a layer of models. Store results on the learners dict."""
         for learner in learners:
-            print(learner)
             learner.fit(dataset, self.holdout_cols)
             self.learners[learner.learner_id] = learner
+
+    def _generate_ensemble_coefficients(
+        self,
+        max_num_models: int,
+        kernel_param: float,
+        ratio_cutoff: float,
+    ) -> np.ndarray:
+        """
+        Generates the weighted ensembled coefficients across all fitted learners.
+
+        :param max_num_models: The maximum number of learners to consider for our weights
+        :param kernel_param: The kernel parameter, amount with which to bias towards strongest
+            performances
+        :param ratio_cutoff: The performance floor which learners must exceed to be considered
+            in the ensemble weights
+        :return: A vector of weighted coefficients aggregated from sufficiently-performing
+            sublearners.
+        """
+
+        # Validate the parameters
+        if ratio_cutoff > 1 or max_num_models < 1:
+            raise InvalidConfigurationError(
+                "The ratio cutoff parameter must be < 1, and max_num_models >= 1, "
+                "otherwise no models will be used for ensembling."
+            )
+
+        learner_ids, coefficients = self._generate_coefficients_matrix()
+
+        # Create weights
+        performances = np.array([self.performances[key] for key in learner_ids])
+        weights = metrics_to_weights(
+            metrics=performances,
+            max_num_models=max_num_models,
+            kernel_param=kernel_param,
+            ratio_cutoff=ratio_cutoff
+        )
+        # Multiply by the coefficients
+        means = coefficients.T.dot(weights)
+
+        return means
+
+    def _generate_coefficients_matrix(self) -> tuple[list[LearnerID], np.ndarray]:
+        """Create the full matrix of learner ids, mapped to its relative coefficients.
+
+        Will result in a m x n matrix, where m = the number of fitted learners in rover
+        and n is the total number of covariates provided.
+
+        Each cell is the i-th learner's coefficient for the j-th covariate, defaulting to 0
+        if the coefficient is not represented in that learner.
+
+        :return:
+        """
+        learner_ids: list[LearnerID] = []
+        performances: np.ndarray = np.array([])
+        for learner_id, learner in self.learners.items():
+
+            if learner.performance and learner.opt_coefs is not None:
+                learner_ids.append(learner_id)
+                performances = np.append(performances, learner.performance)
+
+        # Aggregate the coefficients
+        x_dim, y_dim = len(learner_ids), self.num_covariates
+        coefficients = np.zeros((x_dim, y_dim))
+
+        for row, learner_id in enumerate(learner_ids):
+            opt_coefs = self.learners[learner_id].opt_coefs
+            coefficients[row][list(learner_id)] = opt_coefs
```

Comment:
Here we need to be careful, sometimes, `fixed` covariates are not just intercept, might include other covariates. Probably good to have a `learner_id_to_cov_id` function to handle this.

Reply:
I'd like to hear more, let's discuss at our next meeting